service/controller/resource/subnet/resource.go

```diff
@@ -0,0 +1,273 @@
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToAzureCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, &cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deploymentsClient, err := r.azureClientsFactory.GetDeploymentsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	armTemplate, err := subnet.GetARMTemplate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, allocatedSubnet := range cr.Spec.NetworkSpec.Subnets {
+		deploymentName := fmt.Sprintf("%s-%s", mainDeploymentName, allocatedSubnet.Name)
+		currentDeployment, err := deploymentsClient.Get(ctx, key.ClusterID(&cr), deploymentName)
+		if IsNotFound(err) {
+			// fallthrough
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		parameters, err := r.getDeploymentParameters(ctx, key.ClusterID(&cr), strconv.FormatInt(cr.ObjectMeta.Generation, 10), cr.Spec.NetworkSpec.Vnet.Name, allocatedSubnet)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		desiredDeployment := azureresource.Deployment{
+			Properties: &azureresource.DeploymentProperties{
+				Mode:       azureresource.Incremental,
+				Parameters: key.ToParameters(parameters),
+				Template:   armTemplate,
+			},
+		}
+
+		// We only submit the deployment if it doesn't exist or it exists but it's out of date.
+		shouldSubmitDeployment := currentDeployment.IsHTTPStatus(404)
+		if !shouldSubmitDeployment {
+			shouldSubmitDeployment, err = r.isDeploymentOutOfDate(ctx, cr, currentDeployment)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		if shouldSubmitDeployment {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "template or parameters changed")
+			err = r.createDeployment(ctx, deploymentsClient, key.ClusterID(&cr), deploymentName, desiredDeployment)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+			reconciliationcanceledcontext.SetCanceled(ctx)
+
+			return nil
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "template and parameters unchanged")
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", *currentDeployment.Properties.ProvisioningState))
+
+		if key.IsFailedProvisioningState(*currentDeployment.Properties.ProvisioningState) {
+			r.debugger.LogFailedDeployment(ctx, currentDeployment, err)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "removing failed deployment")
+			_, err = deploymentsClient.Delete(ctx, key.ClusterID(&cr), deploymentName)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) isDeploymentOutOfDate(ctx context.Context, cr capzv1alpha3.AzureCluster, currentDeployment azureresource.DeploymentExtended) (bool, error) {
```

Comment:
This functions decides whether or not the ARM deployment is out of date. For that, we use [the `Generation` field](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta) from the AzureCluster CR. This `Generation` field should change when there is a change in the CR. This means we will submit the ARM template when `AzureCluster` changes, even though there are no changes in the subnets themselves, but I think that's fine.